src/Storages/StorageFilesystem.cpp

```diff
@@ -0,0 +1,306 @@
+namespace fs = std::filesystem;
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int DATABASE_ACCESS_DENIED;
+    extern const int DIRECTORY_DOESNT_EXIST;
+    extern const int LOGICAL_ERROR;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+
+class StorageFilesystemSource final : public ISource
+{
+public:
+    struct PathInfo
+    {
+        ConcurrentBoundedQueue<fs::directory_entry> queue;
+        const String user_files_absolute_path_string;
+        const bool need_check;
+        fs::recursive_directory_iterator directory_iterator;
+        std::atomic_uint32_t rows = 1;
+
+        PathInfo(size_t num_streams, String user_files_absolute_path_string_, bool need_check_)
+            : queue(num_streams * 1000000L)
+            , user_files_absolute_path_string(user_files_absolute_path_string_)
+            , need_check(need_check_)
+            , directory_iterator(user_files_absolute_path_string)
+        {
+        }
+    };
+    using PathInfoPtr = std::shared_ptr<PathInfo>;
+
+    String getName() const override { return DB::String(); }
+
+    Chunk generate() override
+    {
+        size_t current_block_size = 0;
+        std::unordered_map<String, MutableColumnPtr> columns_map;
+        auto names_and_types_in_use = storage_snapshot->getSampleBlockForColumns(columns_in_use).getNamesAndTypesList();
+        for (const auto & [name, type] : names_and_types_in_use)
+        {
+            columns_map[name] = type->createColumn();
+        }
+
+        fs::directory_entry file;
+        while (current_block_size < max_block_size)
+        {
+            if (!path_info->queue.tryPop(file))
+            {
+                LOG_TEST(&Poco::Logger::get("StorageFilesystem"), "No data read from queue, stop processing");
+                break;
+            }
+            current_block_size++;
+            LOG_TEST(&Poco::Logger::get("StorageFilesystem"), "Processing path {} remaining queue size {}",
+                file.path().string(), path_info->queue.size());
+
+            std::error_code ec;
+
+            if (file.is_directory(ec) && !file.is_symlink(ec) && ec.value() == 0)
+            {
+                for (const auto & child : fs::directory_iterator(file, ec))
+                {
+                    fs::path file_path(child);
+                    /// Do not use fs::canonical or fs::weakly_canonical.
+                    /// Otherwise it will not allow to work with symlinks in `user_files_path` directory.
+                    file_path = fs::absolute(file_path).lexically_normal();
+
+                    if (path_info->need_check && file_path.string().find(path_info->user_files_absolute_path_string) != 0)
+                    {
+                        LOG_DEBUG(&Poco::Logger::get("StorageFilesystem"), "Path {} is not inside user_files {}",
+                            file_path.string(),path_info->user_files_absolute_path_string);
+                    }
+                    if (!path_info->queue.push(child))
```

Comment:
Why don't block? I expect it should work with an arbitrary number of files (several billion files as on GitHub).

Reply:
We can't block at least for `max_threads=1`.
I do not like this either, but current implementation will stuck on any number of threads if you have many directories with more directories in them. Something more creative is needed. Or just set 1bln as a limit...

Reply:
We can't block at least for `max_threads=1`.
I do not like this either, but current implementation will stuck on any number of threads if you have many directories with more directories in them. Something more creative is needed. Or just set 1bln as a limit...

Reply:
Hm, we can always use at least one thread reading the results from the queue and forming the result blocks, while other threads will fill the queue. It's ok to use `max_threads + 1` threads, because max_threads is just a hint :) Moreover, only max_threads of them will be IO-heavy, and one extra thread will just fetch the results, so this is ok.

Reply:
I think it will be an elegant solution, let's try it.